### ISSUE TITLE
About mask_sent of mass_step.Code and paper description seem to be inconsistent.

### ISSUE THREAD

**Issue by miaodl** [microsoft/MASS#19]
**trainer.py**

`def mask_sent(self, x, l):
        max_len = 0
        positions, inputs, targets, outputs, = [], [], [], []
        mask_len = round(len(x[:, 0]) * self.params.word_mass)
        len2 = [mask_len for i in range(l.size(0))]

        for i in range(l.size(0)):
            words = x[:l[i], i].tolist()
            start = self.random_start(l[i] - mask_len + 1)

            pos_i, target_i, output_i, input_i = [], [], [], []
            prev_w = None
            for j, w in enumerate(words):
                if j >= start and j < start + mask_len:
                    output_i.append(w) 
                    target_i.append(prev_w)
                    pos_i.append(j - 1)
                    input_i.append(self.mask_word(w))
                else:
                    input_i.append(w)
                prev_w = w

            #################################
            I have read the code and paper.
            In the mass_step, the input of the decoder is "targets".
            
            According to the paper,
            **Should this "target_i[0]" be set to <mask_index>?**
            for example,like this:
            
            **target_i[0]=self.params.mask_index**
            
            #################################

            inputs.append(input_i)
            targets.append(target_i)
            outputs.append(output_i)
            positions.append(pos_i)

        x1  = torch.LongTensor(max(l) , l.size(0)).fill_(self.params.pad_index)
        x2  = torch.LongTensor(mask_len, l.size(0)).fill_(self.params.pad_index)
        y   = torch.LongTensor(mask_len, l.size(0)).fill_(self.params.pad_index)
        pos = torch.LongTensor(mask_len, l.size(0))
        l1  = l.clone()
        l2  = torch.LongTensor(len2)
        for i in range(l.size(0)):
            x1[:l1[i], i].copy_(torch.LongTensor(inputs[i]))
            x2[:l2[i], i].copy_(torch.LongTensor(targets[i]))
            y[:l2[i], i].copy_(torch.LongTensor(outputs[i]))
            pos[:l2[i], i].copy_(torch.LongTensor(positions[i]))
        pred_mask = y != self.params.pad_index
        y = y.masked_select(pred_mask)
        return x1, l1, x2, l2, y, pred_mask, pos`

**Comment by tobyoup** [comment]
Hi @miaodl, the "target_i[0]" should be set to <mask_index> according to the formulation in the paper (https://arxiv.org/pdf/1905.02450.pdf). 
However, for saving training cost in the decoder side (you can check the paragraph "Pre-Training Details" in Section 4.1 of the paper), we remove the 
the masked tokens but keep the positional embedding of the unmasked tokens unchanged. So for ease of implementation and unification, we set "target_i[0]" as the last ground truth token.
